File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        memo = {}
        start = 0
        max_len = 0
        for end in range(len(s)):
            if s[end] in memo:
                start = max(start, memo[s[end]] + 1)
            max_len = max(max_len, (end - start + 1))
            memo[s[end]] = end

        return max_len
        

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashMap<Character, Integer> memo = new HashMap<>();
        int maxLength = 0;
        int start = 0;
        for(int end = 0; end < s.length(); end++){
            char current = s.charAt(end);
            if (memo.containsKey(current)) {
                start = Math.max(start, memo.get(current) + 1);
            }
            maxLength = Math.max(maxLength, (end - start) + 1);
            memo.put(current, end);
        }
        return maxLength;
    }
}